controlplane/kubeadm/controllers/kubeadm_control_plane_controller_test.go

```diff
@@ -55,26 +58,34 @@ func (tc *templateCloner) CloneTemplate(_ context.Context, _ client.Client, ref
 
 type kubeadmConfigGenerator struct{}
 
-func (kcg *kubeadmConfigGenerator) GenerateKubeadmConfig(_ context.Context, _ client.Client, namespace, _, _ string, _ *bootstrapv1.KubeadmConfigSpec, _ *metav1.OwnerReference) (*corev1.ObjectReference, error) {
+func (kcg *kubeadmConfigGenerator) GenerateKubeadmConfig(_ context.Context, _ client.Client, namespace, _, _ string, bootstrapConfig *bootstrapv1.KubeadmConfigSpec, _ *metav1.OwnerReference) (*corev1.ObjectReference, error) {
+	name := "joinConfig"
+
+	if bootstrapConfig.JoinConfiguration == nil {
+		// This is a init configuration
+		name = "initConfig"
+	}
+
 	result := &corev1.ObjectReference{
 		Kind:       "KubeadmConfig",
 		APIVersion: bootstrapv1.GroupVersion.String(),
 		Namespace:  namespace,
-		Name:       "generatedKubeadmConfig",
+		Name:       name,
 	}
 	return result, nil
 }
 
 type machineGenerator struct{}
 
-func (mg *machineGenerator) GenerateMachine(ctx context.Context, c client.Client, namespace, _, _, _ string, infraRef, bootstrapRef *corev1.ObjectReference, _ map[string]string, _ *metav1.OwnerReference) error {
+func (mg *machineGenerator) GenerateMachine(ctx context.Context, c client.Client, namespace, namePrefix, _, _ string, infraRef, bootstrapRef *corev1.ObjectReference, _ map[string]string, _ *metav1.OwnerReference) error {
```

Comment:
Something feels "off" to me about this abstraction - the logic is quite similar to the real implementation, except it does client-side name generation and doesn't handle setting the owner reference (if present). And in `TestCloneConfigsAndGenerateMachine`, we're validating a lot of the fields in the generated machines, but these fields are all being populated in this test implementation, so we're validating that the test implementation works as expected... We don't have to fix it in this PR, but it would be nice if we could eliminate or alter the test-specific code so we're doing more validation of the actual code.

Reply:
Yeah, this was a bit tricky, but the answer might actually be to break things out a bit more. I ended up having to test the values a bit to make sure they were getting passed into the generate method. I'll work to see if I can refactor things a bit more cleanly.

Reply:
Resolved as a separate PR in https://github.com/kubernetes-sigs/cluster-api/pull/1991